### PR TITLE
Allow override of name and description of McpClientTool

### PR DESCRIPTION
**Pull request by PederHP**
Addresses #154 by providing utility to clone McpClientTool with a new name or description. This will mostly be useful for hosts that need to be robust against user-configurable MCP servers (to avoid name conflicts by prefixing with the server name) and for hosts that use MCP servers known at development time where evaluation pipelines have shown that the name and/or descriptions of tools do not result in acceptable tool calling accuracy.

This could be used like this:
```csharp
if (myExamples.ContainsKey(tool.Name))
{
  tool = tool.WithDescription(tool.Description + "\n\nExample Usage:\n" + myExamples[tool.Name]);
}
```

or

```csharp
return tools.Select(tool => tool.WithName($"{client.ServerInfo.Name}_{tool.Name}"));
```

### CODE REVIEW

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 1]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
...
+    private readonly string? _nameOverride;
+    private readonly string? _descriptionOverride;
 
-    internal McpClientTool(IMcpClient client, Tool tool)
+    internal McpClientTool(IMcpClient client, Tool tool, string? nameOverride = null, string? descriptionOverride = null)
     {
         _client = client;
         ProtocolTool = tool;
+        _nameOverride = nameOverride;
+        _descriptionOverride = descriptionOverride;
```
Comment (halter73) [inline src/ModelContextProtocol/Client/McpClientTool.cs:20]: If we're adding the fields anyway, wouldn't it be simpler to do it like so?

```suggestion
        _name = nameOverride ?? tool.Name ?? string.Empty;
        _description = descriptionOverride ?? ProtocolTool.Description ?? string.Empty;
```

And then just use `=> _name;` and `=> _description;` as the getters?

I'd also want to consider what's so bad about just making these properties publicly settable. I know it's not exactly pure, but we don't have to make any promises about what happens if you set it. Or we could promise to ignore whatever you set which is essentially what we're doing with this.

Comment (PederHP) [reply src/ModelContextProtocol/Client/McpClientTool.cs:20]: I wouldn't mind public setters personally, but I had the impression MEAI had immutable AIFunctions and thus wanted to align with that.

Comment (stephentoub) [reply src/ModelContextProtocol/Client/McpClientTool.cs:20]: Making these mutable seems reasonable.

Comment (halter73) [reply src/ModelContextProtocol/Client/McpClientTool.cs]: AIFunction has no WithName or WithDescription members to override either. You need to have a McpClientTool to use it. Isn't it the same with a normal setter?

Comment (stephentoub) [reply src/ModelContextProtocol/Client/McpClientTool.cs]: I don't think so :(

Comment (PederHP) [reply src/ModelContextProtocol/Client/McpClientTool.cs]: > AIFunction has no WithName or WithDescription members to override either. You need to have a McpClientTool to use it. Isn't it the same with a normal setter?

But that's ok - this is for MCP host developers who need to adjust details of McpClientTools before passing them on as AIFunctions. I think making this a slightly awkward operation is ok - it should be used for a reason. The old C++ `reinterpret_cast<T>` pattern of ensuring someone who uses this feature actually needs it, without making it too annoying.

Because while I seems like an exotic use case - when one does hit it, it's crippling not to be able to override these properties - and having to write a carbon copy of McpClientTool and the handler extensions isn't great ergonomics.

However, I am certainly open to something that ends up looking more idiomatic (and I actually think being to override a virtual getter while also providing a new setter should be supported by the language - because the alternative is gunky stuff like this - but I guess there are permutations where the setter is hidden because of access modifiers that make this problematic).

**Review by halter73 — APPROVED** [review]
While I'm not a huge fan of "With" methods outside of builder types, I don't have a better alternative at the moment😄

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 2]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
...
+    }
+
+    /// <summary>
+    /// Creates a new instance of the tool with the specified name.
+    /// This is useful for optimizing the tool name for specific models or for prefixing the tool name with a (usually server-derived) namespace to avoid conflicts.
+    /// The server will still be called with the original tool name, so no mapping is required.
+    /// </summary>
+    /// <param name="name">The model-facing name to give the tool. Pass null to clear the name override and to use the MCP Tool name again.</param>
+    /// <returns>Copy of this McpClientTool with the provided name</returns>
+    public McpClientTool WithName(string? name)
```
Comment (halter73) [inline src/ModelContextProtocol/Client/McpClientTool.cs:30]: I know I approved this, by why allow null names here? Shouldn't it just throw in that case?

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 3]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
...
+    {
+        return new McpClientTool(_client, ProtocolTool, _name, description);
     }
 
     /// <summary>Gets the protocol <see cref="Tool"/> type for this instance.</summary>
     public Tool ProtocolTool { get; }
 
     /// <inheritdoc/>
-    public override string Name => ProtocolTool.Name;
+    public override string Name => _name!;
```
Comment (halter73) [inline src/ModelContextProtocol/Client/McpClientTool.cs:52]:
```suggestion
    public override string Name => _name ?? string.Empty;
```

If we're really sure it cannot be null, we should still have something like `_name ?? throw new UnreachableException();`.

File: src/ModelContextProtocol/Client/McpClientTool.cs (hunk 2 above)
Comment (stephentoub) [reply src/ModelContextProtocol/Client/McpClientTool.cs:30]: I agree this should prohibit null. I'm not seeing a use case for wanting to revert back to the original name, and it doesn't actually let you revert back to the previous name, just the original.

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 4]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
...
+
+    /// <summary>
+    /// Creates a new instance of the tool with the specified description.
+    /// This can be used to provide modified or additional (e.g. examples) context to the model about the tool.
+    /// This will in general require a hard-coded mapping in the client. 
+    /// It is not recommended to use this without running evaluations to ensure the model actually benefits from the custom description.
+    /// </summary>
+    /// <param name="description">The description to give the tool. Pass null to clear the description override and to use the MCP Tool description again.</param>
+    /// <returns>Copy of this McpClientTool with the provided description</returns>
+    public McpClientTool WithDescription(string? description)
```
Comment (stephentoub) [inline src/ModelContextProtocol/Client/McpClientTool.cs:43]: Please add some tests as well. Thanks!

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 5]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
...
+    /// <summary>
+    /// Creates a new instance of the tool with the specified name.
+    /// This is useful for optimizing the tool name for specific models or for prefixing the tool name with a (usually server-derived) namespace to avoid conflicts.
+    /// The server will still be called with the original tool name, so no mapping is required.
+    /// </summary>
+    /// <param name="name">The model-facing name to give the tool.</param>
+    /// <returns>Copy of this McpClientTool with the provided name</returns>
+    public McpClientTool WithName(string name)
+    {
+        return new McpClientTool(_client, ProtocolTool, name, _description);
```
Comment (eiriktsarpalis) [inline src/ModelContextProtocol/Client/McpClientTool.cs:32]: Should add runtime validation in public APIs when the annotation is non-nullable.
```suggestion
        Throw.IfNull(name);
        return new McpClientTool(_client, ProtocolTool, name, _description);
```

File: src/ModelContextProtocol/Client/McpClientTool.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/ModelContextProtocol/Client/McpClientTool.cs:52]: I think we should remove nullability from the fields altogether and have the constructor normalize.

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 6]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
...
+
+    /// <summary>
+    /// Creates a new instance of the tool with the specified description.
+    /// This can be used to provide modified or additional (e.g. examples) context to the model about the tool.
+    /// This will in general require a hard-coded mapping in the client. 
+    /// It is not recommended to use this without running evaluations to ensure the model actually benefits from the custom description.
+    /// </summary>
+    /// <param name="description">The description to give the tool.</param>
+    /// <returns>Copy of this McpClientTool with the provided description</returns>
+    public McpClientTool WithDescription(string description)
```
Comment (eiriktsarpalis) [inline src/ModelContextProtocol/Client/McpClientTool.cs:43]: Could you also add a couple of unit tests validating the methods (including ensuring that null parameters throw?)

File: src/ModelContextProtocol/Client/McpClientTool.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/ModelContextProtocol/Client/McpClientTool.cs:52]: > _name ?? string.Empty;

I suspect defaulting to an empty string for tool names is likely going to be a source of problems (not the case for descriptions though). Do we need to reject names that are null or whitespace altogether?

File: src/ModelContextProtocol/Client/McpClientTool.cs [hunk 7]
```diff
@@ -9,21 +9,50 @@ namespace ModelContextProtocol.Client;
 public sealed class McpClientTool : AIFunction
 {
     private readonly IMcpClient _client;
+    private readonly string? _name;
+    private readonly string? _description;
```
Comment (eiriktsarpalis) [inline src/ModelContextProtocol/Client/McpClientTool.cs:13]:
```suggestion
    private readonly string _name;
    private readonly string _description;
```

**Comment by eiriktsarpalis** [comment]
> While I'm not a huge fan of "With" methods outside of builder types, I don't have a better alternative at the moment😄

I think using `With` is perfectly fine in the context of immutable persistable updates, C# records use the `with` keyword for essentially the same thing.

File: src/ModelContextProtocol/Client/McpClientTool.cs (hunk 3 above)
Comment (PederHP) [reply src/ModelContextProtocol/Client/McpClientTool.cs:52]: Whitespace tool names might be valid, but it's a really strange thing to do. `null` is always an error as then they can't really be called by the LLM or even invoked server-side.

File: tests/ModelContextProtocol.Tests/Client/McpClientExtensionsTests.cs [hunk 8]
```diff
@@ -171,4 +171,32 @@ public async Task GetPromptsAsync_HonorsJsonSerializerOptions()
...
+
+    [Fact]
+    public async Task WithName_ChangesToolName()
+    {
+        JsonSerializerOptions options = new(JsonSerializerOptions.Default);
+        IMcpClient client = await CreateMcpClientForServer();
+
+        var tool = (await client.ListToolsAsync(options, TestContext.Current.CancellationToken)).FirstOrDefault();
+        var originalName = tool?.Name;
+        var renamedTool = tool?.WithName("RenamedTool");
```
Comment (eiriktsarpalis) [inline tests/ModelContextProtocol.Tests/Client/McpClientExtensionsTests.cs:183]: Do we expect the client fixture to not return tools?
```suggestion
        var tool = (await client.ListToolsAsync(options, TestContext.Current.CancellationToken)).First();
        var originalName = tool.Name;
        var renamedTool = tool.WithName("RenamedTool");
```
